### DISCUSSION TITLE
`assert_type` false positive comparing function and callable protocol

### DISCUSSION THREAD

**Discussion by bedlamzd** [microsoft/pyright#7350] [Q&A]
I'm not sure if this is a bug or I'm missing something but the following fails

```python

class F(Protocol):
    def __call__(self, foo: str = ...) -> bool: ...


def f(foo: str = "default") -> bool:
    return bool(foo)


assert_type(f, F)
# error: "assert_type" mismatch: expected "F" but received "(foo: str = "default") -> bool" (reportAssertTypeFailure)
```

mypy for the same code doesn't produce any errors

**Comment by erictraut** [comment]
I think pyright is behaving correctly here. The `assert_type` assertion checks for type equivalence between the two types. While function `f` is a subtype of protocol `F`, the opposite is not true. Functions defined with a `def` statement have other properties that are not in the protocol, such as `__name__`.

**Reply by bedlamzd** [reply]
Fair enough. Then do you have any suggestions on how to check the following?

```python
def b(*, bar: str = "default") -> bool:
    return bool(bar)


assert_type(b, Callable[[str], bool])
# error: "assert_type" mismatch: expected "(str) -> bool" but received "(*, bar: str = "default") -> bool" (reportAssertTypeFailure)
```

This is in essence my original problem. I cannot express keyword-only params using `Callable`, unpacking `TypedDict` doesn't seem to be supported and it seems I cannot use `Protocol`?

For the context, I had an idea to write a test that way to confirm method signature is preserved after applying decorators. You can take a look [there](https://github.com/roman-right/beanie/pull/886#issuecomment-1965655632) if you're curious. Although the project uses mypy, I'd like to know how to do it with pyright as well, since that's what I usually use.

**Reply by erictraut** [reply]
You could check for type consistency rather than type equivalence by replacing the `assert_type` with a simple assignment.

```python
test_f: F = f
```

**Reply by bedlamzd** [reply]
Right, didn't cross my mind for some reason. Thank you!
